src/dmd/toir.d

```diff
@@ -774,8 +774,18 @@ void setClosureVarOffset(FuncDeclaration fd)
 void buildClosure(FuncDeclaration fd, IRState *irs)
 {
     //printf("buildClosure(fd = %s)\n", fd.toChars());
+    const oldValue = fd.requiresClosure;
     if (fd.needsClosure())
     {
+        if (oldValue != fd.requiresClosure)
+        {
+            /* https://issues.dlang.org/show_bug.cgi?id=23112
+             * This can shift due to templates being expanded that access alias symbols.
+             * Always make it an error.
+             */
+            fd.checkClosure();   // give decent diagnostic
+        }
```

Comment:
Can't this be in semantic3 instead?

Reply:
Oh, I wish. But I can't think of a way.

Reply:
Perhaps a fourth pass to tackle heuristics once the full module unit is known? This'll apply for many other errors or fixups that currently occur during codegen (when they should be handled earlier).

Can't say I'm too keen on that though as these still would succeed `__traits(compiles)` tests.

Reply:
Don't we have the `Module.runDeferredSemantic3` just for that?

Reply:
I suspect a 4th pass would lead to more bugs than it fixes.